core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala

```diff
@@ -118,14 +118,14 @@ private[spark] class TaskResultGetter(sparkEnv: SparkEnv, scheduler: TaskSchedul
 
   def enqueueFailedTask(taskSetManager: TaskSetManager, tid: Long, taskState: TaskState,
     serializedData: ByteBuffer) {
-    var reason : TaskEndReason = UnknownReason
+    var reason : TaskFailedReason = UnknownReason
     try {
       getTaskResultExecutor.execute(new Runnable {
         override def run(): Unit = Utils.logUncaughtExceptions {
           val loader = Utils.getContextOrSparkClassLoader
           try {
             if (serializedData != null && serializedData.limit() > 0) {
-              reason = serializer.get().deserialize[TaskEndReason](
+              reason = serializer.get().deserialize[TaskFailedReason](
```

Comment:
Good question.  its sent as an executor status update in various failure handling scenarios here, though its not so nicely grouped that its super-obvious they always go together, unfortunately:

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L357

I walked through the various cases and convinced myself they went together.  I made one more minor improvement to make that more clear, a couple of `.toTaskEndReason` methods could be renamed to `.toTaskFailedReason` (https://github.com/apache/spark/pull/15181/commits/d991a3c575251f787beb215820fbae9df2cbf99f).

Really, though there were two other things that convinced me this was OK:
1) the only other possible thing a `TaskEndReason` could be is `Success`.  But you'll notice that `Success` is never serialized as the msg -- its just implicit with `TaskState.FINISHED`, and then the `Success` part just gets dropped in on the driver-side here: https://github.com/apache/spark/blob/248922fd4fb7c11a40304431e8cc667a8911a906/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L669

2) The TaskEndReason was already getting blind-casted before anyway: https://github.com/apache/spark/blob/248922fd4fb7c11a40304431e8cc667a8911a906/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L710
This was already in the same thread, without any try/ catch etc. checks anyway.  Well, except for this: https://github.com/apache/spark/blob/248922fd4fb7c11a40304431e8cc667a8911a906/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L701

which I decided wasn't a concern mostly b/c of reason (1) again.

There's more cleanup we could do here:
* `TaskState.isFailed` is unused, instead there is a different hard-coded check here: https://github.com/apache/spark/blob/248922fd4fb7c11a40304431e8cc667a8911a906/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L376
* We could more directly fix the weak coupling between TaskState and TaskEndReason by replacing `ExecutorBackend.statusUpdate` with more particular methods, so that the coupling is more explicit.

and probably other related things too.  If you have a strong preference, I could address those here, but figured it was best to just get in this small cleanup I felt confident in for now.